storage/remote/codec.go

```diff
@@ -183,6 +183,150 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
 	}
 }
 
+func StreamChunkedReadResponses(
+	stream io.Writer,
+	queryIndex int64,
+	ss storage.SeriesSet,
+	sortedExternalLabels []prompb.Label,
+	sampleLimit int,
+	maxChunksInFrame int,
+) error {
+	var numSamples = 0
+
+	for ss.Next() {
+		series := ss.At()
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		for iter.Next() {
+			// TODO(bwplotka): Expose maxChunks as flag?
+			chks, samples, err := encodeChunks(iter, 120, maxChunksInFrame, sampleLimit-numSamples)
+			if err != nil {
+				return err
+			}
+			numSamples += samples
+			if len(chks) == 0 {
+				break
+			}
+
+			b, err := proto.Marshal(&prompb.ChunkedReadResponse{
+				// TODO(bwplotka): Do we really need multiple?
+				ChunkedSeries: []*prompb.ChunkedSeries{
+					{
+						// TODO(bwplotka): Oneof? We don't need to send labels all the time.
+						Labels: lbls,
+						Chunks: chks,
+					},
+				},
+				QueryIndex: queryIndex,
+			})
+			if err != nil {
+				return errors.Wrap(err, "marshal ChunkedReadResponse")
+			}
+
+			if _, err := stream.Write(b); err != nil {
+				return errors.Wrap(err, "write to stream")
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return err
+		}
+	}
+	if err := ss.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, allowedNumSamples int) ([]prompb.Chunk, int, error) {
```

Comment:
If I read this correctly, this is iterating over the samples in the series and rebuilding the chunks.  Can we extend TSDB to just give us the chunks and save the hassle?

Reply:
ps I want this for Cortex too, so I did it: https://github.com/cortexproject/cortex/pull/1427/files#diff-a6d94f56da1ed83021ef693d2dd12b55R58

Reply:
I gather this PR is just the first step.

Reply:
Yes. I would really really vote for avoiding scope creep and merge working bits by bits.

We can add avoid re-encoding easily in other PRs, it's just internal optimization. I promise I will do it, but let's ship massive win with even this implementation. ;p

Reply:
Happy to do this optimisation as another PR.  Even happy to do that myself.

Reply:
to link: https://github.com/prometheus/tsdb/pull/659

But I think iterator would be awesome instead.